src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/custom.py

```diff
@@ -83,6 +84,34 @@ def which(binary):
     return None
 
 
+def register_providers(cli_ctx):
+    """Registers the required Azure resource providers for AKS, if necessary."""
```

Comment:
msrestazure is supposed to capture such cases and register the provider for you automatically. the logic is [here](https://github.com/Azure/msrestazure-for-python/blob/master/msrestazure/tools.py#L41-L62)
Let me take a look at this and figure out why, also we have a work item for leveraging a new ARM error field to detect such case reliably, so i figure now is the good time to get it done.

Reply:
@troydai @mboersma, l let us not merge this PR till i sort out the root cause today.

Reply:
I thought this was supposed to work automatically as well. The acs module actually had similar code many months ago, but I was urged to take it out when the core module made this behavior automatic.

I can see that auto-registration isn't working correctly right now:
```console
$ az aks create -n foo -g bar
Operation failed with status: 'Bad Request'. Details: Service principal clientID: http://clitestgpccorvewl not found in Active Directory tenant 41bb1bd9-8413-453b-9de5-1937fef090b3, Please see https://aka.ms/acs-sp-help for more details.
$ az provider show -n Microsoft.ContainerService -o table
Namespace                   RegistrationState
--------------------------  -------------------
Microsoft.ContainerService  Unregistered
```

Reply:
@mboersma , are you able to reproduce this issue? If yes, can you send me the error response payload? I just now tried out in a new subscription where the AKS is not registered, running `az aks create` did handle the registration for you:
```bash
(env) C:\SDK\azure-cli>az aks create -g yg -n yg-aks
Resource provider 'Microsoft.ContainerService' used by this operation is not registered. We are registering for you.
Registration succeeded.
...
```

Reply:
Actually, my example is wrong, and you're right--I do see the auto-registration happening for `Microsoft.ContainerService`. But I don't see it for the other required providers, presumably because the CLI isn't calling any of their API endpoints directly? Here's a way to reproduce it:

```console
$ az provider unregister -n Microsoft.Compute
$ az aks create -n foo -g bar
```

Reply:
can you share the error response payload using "--debug"? I like to see whether `Microsoft.compute` gets mentioned or not.

Reply:
```console
$ az provider list -o table | head
Namespace                                 RegistrationState
----------------------------------------  -------------------
Microsoft.Compute                         Unregistered
Microsoft.ContainerService                Unregistered
Microsoft.Network                         Unregistered
Microsoft.Storage                         Unregistered
84codes.CloudAMQP                         NotRegistered
AppDynamics.APM                           NotRegistered
Aspera.Transfers                          NotRegistered
$ az aks create -n foobar -g bar --debug
Command arguments: ['aks', 'create', '-n', 'foobar', '-g', 'bar', '--debug']
Event: Cli.PreExecute []
Event: CommandParser.OnGlobalArgumentsCreate [<function CLILogging.on_global_arguments at 0x10eb5f840>, <function OutputProducer.on_global_arguments at 0x10ec02378>, <function CLIQuery.on_global_arguments at 0x10ec3be18>]
Event: CommandInvoker.OnPreCommandTableCreate [<function CLILogging.remove_logger_flags at 0x10eb5f8c8>]
Installed command modules ['servicefabric', 'dls', 'appservice', 'dla', 'storage', 'cloud', 'container', 'cognitiveservices', 'batch', 'eventgrid', 'network', 'vm', 'ams', 'feedback', 'acr', 'iot', 'monitor', 'acs', 'rdbms', 'interactive', 'servicebus', 'backup', 'lab', 'advisor', 'reservations', 'find', 'keyvault', 'eventhubs', 'role', 'billing', 'redis', 'resource', 'cdn', 'consumption', 'configure', 'extension', 'cosmosdb', 'batchai', 'sql', 'profile', 'dms']
Loaded module 'servicefabric' in 0.001 seconds.
Loaded module 'dls' in 0.003 seconds.
Loaded module 'appservice' in 0.003 seconds.
Loaded module 'dla' in 0.003 seconds.
Event: CommandLoader.OnLoadCommandTable []
Loaded module 'storage' in 0.017 seconds.
Loaded module 'cloud' in 0.001 seconds.
Loaded module 'container' in 0.001 seconds.
Loaded module 'cognitiveservices' in 0.001 seconds.
Event: CommandLoader.OnLoadCommandTable []
Loaded module 'batch' in 0.006 seconds.
Loaded module 'eventgrid' in 0.001 seconds.
Loaded module 'network' in 0.015 seconds.
Loaded module 'vm' in 0.009 seconds.
Event: CommandLoader.OnLoadCommandTable []
Loaded module 'ams' in 0.003 seconds.
Loaded module 'feedback' in 0.001 seconds.
Loaded module 'acr' in 0.003 seconds.
Loaded module 'iot' in 0.003 seconds.
Loaded module 'monitor' in 0.004 seconds.
Loaded module 'acs' in 0.003 seconds.
Loaded module 'rdbms' in 0.003 seconds.
Loaded module 'interactive' in 0.001 seconds.
Loaded module 'servicebus' in 0.004 seconds.
Loaded module 'backup' in 0.012 seconds.
Loaded module 'lab' in 0.003 seconds.
Loaded module 'advisor' in 0.002 seconds.
Loaded module 'reservations' in 0.002 seconds.
Loaded module 'find' in 0.001 seconds.
Loaded module 'keyvault' in 0.004 seconds.
Loaded module 'eventhubs' in 0.003 seconds.
Loaded module 'role' in 0.003 seconds.
Loaded module 'billing' in 0.002 seconds.
Loaded module 'redis' in 0.002 seconds.
Loaded module 'resource' in 0.004 seconds.
Event: CommandLoader.OnLoadCommandTable []
Loaded module 'cdn' in 0.003 seconds.
Loaded module 'consumption' in 0.005 seconds.
Loaded module 'configure' in 0.001 seconds.
Loaded module 'extension' in 0.001 seconds.
Loaded module 'cosmosdb' in 0.003 seconds.
Loaded module 'batchai' in 0.002 seconds.
Loaded module 'sql' in 0.004 seconds.
Loaded module 'profile' in 0.003 seconds.
Loaded module 'dms' in 0.002 seconds.
Loaded all modules in 0.148 seconds. (note: there's always an overhead with the first module loaded)
Extensions directory: '/Users/matt/.azure/cliextensions'
Event: CommandInvoker.OnPreCommandTableTruncate []
Event: CommandInvoker.OnPostCommandTableCreate [<function add_id_parameters at 0x10ee582f0>]
Event: CommandInvoker.OnCommandTableLoaded []
Event: CommandInvoker.OnPreParseArgs [<function _documentdb_deprecate at 0x10f5bff28>]
Event: CommandInvoker.OnPostParseArgs [<function OutputProducer.handle_output_argument at 0x10ec02400>, <function CLIQuery.handle_query_parameter at 0x10ec3bea0>, <function handler at 0x10f33a378>]
Use existing SSH public key file: /Users/matt/.ssh/id_rsa.pub
Getting management service client client_type=ContainerServiceClient
Getting management service client client_type=ResourceManagementClient
attempting to read file /Users/matt/.azure/accessTokens.json as utf-8-sig
adal-python : 861ccaf2-eb64-4485-823a-48eecab0ffe1 - Authority:Performing instance discovery: ...
adal-python : 861ccaf2-eb64-4485-823a-48eecab0ffe1 - Authority:Performing static instance discovery
adal-python : 861ccaf2-eb64-4485-823a-48eecab0ffe1 - Authority:Authority validated via static instance discovery
adal-python : 861ccaf2-eb64-4485-823a-48eecab0ffe1 - TokenRequest:Getting token from cache with refresh if necessary.
adal-python : 861ccaf2-eb64-4485-823a-48eecab0ffe1 - CacheDriver:finding with query keys: {'_clientId': '...', 'userId': '...'}
adal-python : 861ccaf2-eb64-4485-823a-48eecab0ffe1 - CacheDriver:Looking for potential cache entries: {'_clientId': '...', 'userId': '...'}
adal-python : 861ccaf2-eb64-4485-823a-48eecab0ffe1 - CacheDriver:Found 3 potential entries.
adal-python : 861ccaf2-eb64-4485-823a-48eecab0ffe1 - CacheDriver:Resource specific token found.
adal-python : 861ccaf2-eb64-4485-823a-48eecab0ffe1 - CacheDriver:Returning token from cache lookup, AccessTokenId: b'Jpg3my4ZCs0ggk6y5CPNXKHoZYObX8zGQVjsUN8eSS4=', RefreshTokenId: b'BQDDxlHThxapPpBWJohyeBJ3ko4dVUaGuDkvEl/Jm/s='
msrest.pipeline : Configuring request: timeout=100, verify=True, cert=None
msrest.pipeline : Configuring proxies: ''
msrest.pipeline : Evaluate proxies against ENV settings: True
msrest.pipeline : Configuring redirects: allow=True, max=30
msrest.pipeline : Configuring retry: max_retries=4, backoff_factor=0.8, max_backoff=90
urllib3.connectionpool : Starting new HTTPS connection (1): management.azure.com
urllib3.connectionpool : https://management.azure.com:443 "GET /subscriptions/12345678-afaa-4874-a199-33424d31b95d/resourcegroups/bar?api-version=2017-05-10 HTTP/1.1" 200 255
msrest.http_logger : Request URL: 'https://management.azure.com/subscriptions/12345678-afaa-4874-a199-33424d31b95d/resourcegroups/bar?api-version=2017-05-10'
msrest.http_logger : Request method: 'GET'
msrest.http_logger : Request headers:
msrest.http_logger :     'User-Agent': 'python/3.6.5 (Darwin-17.5.0-x86_64-i386-64bit) requests/2.18.4 msrest/0.4.29 msrest_azure/0.4.29 resourcemanagementclient/1.2.1 Azure-SDK-For-Python AZURECLI/2.0.33'
msrest.http_logger :     'Accept-Encoding': 'gzip, deflate'
msrest.http_logger :     'Accept': 'application/json'
msrest.http_logger :     'Connection': 'keep-alive'
msrest.http_logger :     'Authorization': '*****'
msrest.http_logger :     'x-ms-client-request-id': '34b9a208-553c-11e8-80ed-784f436a9564'
msrest.http_logger :     'CommandName': 'aks create'
msrest.http_logger :     'Content-Type': 'application/json; charset=utf-8'
msrest.http_logger :     'accept-language': 'en-US'
msrest.http_logger : Request body:
msrest.http_logger : None
msrest.http_logger : Response status: 200
msrest.http_logger : Response headers:
msrest.http_logger :     'Cache-Control': 'no-cache'
msrest.http_logger :     'Pragma': 'no-cache'
msrest.http_logger :     'Content-Type': 'application/json; charset=utf-8'
msrest.http_logger :     'Content-Encoding': 'gzip'
msrest.http_logger :     'Expires': '-1'
msrest.http_logger :     'Vary': 'Accept-Encoding'
msrest.http_logger :     'x-ms-ratelimit-remaining-subscription-reads': '14999'
msrest.http_logger :     'x-ms-request-id': 'd7acd991-6a1f-4ac2-9bdf-cfeea5d98dfd'
msrest.http_logger :     'x-ms-correlation-request-id': 'd7acd991-6a1f-4ac2-9bdf-cfeea5d98dfd'
msrest.http_logger :     'x-ms-routing-request-id': 'WESTUS2:20180511T165611Z:d7acd991-6a1f-4ac2-9bdf-cfeea5d98dfd'
msrest.http_logger :     'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'
msrest.http_logger :     'X-Content-Type-Options': 'nosniff'
msrest.http_logger :     'Date': 'Fri, 11 May 2018 16:56:11 GMT'
msrest.http_logger :     'Content-Length': '255'
msrest.http_logger : Response content:
msrest.http_logger : b'{"id":"/subscriptions/12345678-afaa-4874-a199-33424d31b95d/resourceGroups/bar","name":"bar","location":"eastus","properties":{"provisioningState":"Succeeded"}}'
adal-python : 23e9127d-7111-4c2d-9c50-1d4a829d84a0 - Authority:Performing instance discovery: ...
adal-python : 23e9127d-7111-4c2d-9c50-1d4a829d84a0 - Authority:Performing static instance discovery
adal-python : 23e9127d-7111-4c2d-9c50-1d4a829d84a0 - Authority:Authority validated via static instance discovery
adal-python : 23e9127d-7111-4c2d-9c50-1d4a829d84a0 - TokenRequest:Getting token from cache with refresh if necessary.
adal-python : 23e9127d-7111-4c2d-9c50-1d4a829d84a0 - CacheDriver:finding with query keys: {'_clientId': '...', 'userId': '...'}
adal-python : 23e9127d-7111-4c2d-9c50-1d4a829d84a0 - CacheDriver:Looking for potential cache entries: {'_clientId': '...', 'userId': '...'}
adal-python : 23e9127d-7111-4c2d-9c50-1d4a829d84a0 - CacheDriver:Found 3 potential entries.
adal-python : 23e9127d-7111-4c2d-9c50-1d4a829d84a0 - CacheDriver:Resource specific token found.
adal-python : 23e9127d-7111-4c2d-9c50-1d4a829d84a0 - CacheDriver:Returning token from cache lookup, AccessTokenId: b'Jpg3my4ZCs0ggk6y5CPNXKHoZYObX8zGQVjsUN8eSS4=', RefreshTokenId: b'BQDDxlHThxapPpBWJohyeBJ3ko4dVUaGuDkvEl/Jm/s='
msrest.pipeline : Configuring request: timeout=100, verify=True, cert=None
msrest.pipeline : Configuring proxies: ''
msrest.pipeline : Evaluate proxies against ENV settings: True
msrest.pipeline : Configuring redirects: allow=True, max=30
msrest.pipeline : Configuring retry: max_retries=4, backoff_factor=0.8, max_backoff=90
urllib3.connectionpool : Starting new HTTPS connection (1): management.azure.com
urllib3.connectionpool : https://management.azure.com:443 "PUT /subscriptions/12345678-afaa-4874-a199-33424d31b95d/resourceGroups/bar/providers/Microsoft.ContainerService/managedClusters/foobar?api-version=2017-08-31 HTTP/1.1" 400 146
msrest.http_logger : Request URL: 'https://management.azure.com/subscriptions/12345678-afaa-4874-a199-33424d31b95d/resourceGroups/bar/providers/Microsoft.ContainerService/managedClusters/foobar?api-version=2017-08-31'
msrest.http_logger : Request method: 'PUT'
msrest.http_logger : Request headers:
msrest.http_logger :     'User-Agent': 'python/3.6.5 (Darwin-17.5.0-x86_64-i386-64bit) requests/2.18.4 msrest/0.4.29 msrest_azure/0.4.29 azure-mgmt-containerservice/3.0.1 Azure-SDK-For-Python AZURECLI/2.0.33'
msrest.http_logger :     'Accept-Encoding': 'gzip, deflate'
msrest.http_logger :     'Accept': 'application/json'
msrest.http_logger :     'Connection': 'keep-alive'
msrest.http_logger :     'Authorization': '*****'
msrest.http_logger :     'x-ms-client-request-id': '34b9a208-553c-11e8-80ed-784f436a9564'
msrest.http_logger :     'CommandName': 'aks create'
msrest.http_logger :     'Content-Type': 'application/json; charset=utf-8'
msrest.http_logger :     'accept-language': 'en-US'
msrest.http_logger :     'Content-Length': '1231'
msrest.http_logger : Request body:
msrest.http_logger : {"location": "eastus", "properties": {"dnsPrefix": "foobar-bar-5bff7a", "kubernetesVersion": "", "agentPoolProfiles": [{"name": "nodepool1", "count": 3, "vmSize": "Standard_DS1_v2", "dnsPrefix": "foobar-bar-5bff7a", "storageProfile": "ManagedDisks", "osType": "Linux"}], "linuxProfile": {"adminUsername": "azureuser", "ssh": {"publicKeys": [{"keyData": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQD2UsLB0WOugWCUN7MFm6CgJdXQFtG+n7uUvffTMBnr2dfDt0b5pxwobrhA56FTDamGl+AIVhtsCe58qihWCLK239zfdHyB5s4eAaccuKbzYu/ZB82JfuZp1jz4q/O71obY+EtHg5rW2/3EnANll95Ptn60B+O1G3NvjL9g9l+E/Lcn7pGAmeVgU8GLUp4pZ6iV2lE8q1+Uev/Ih7kvl6zFD+akX4I6sCd4OUs2ckuFPM9g0Zsbzp+lnNiJ/kTFA3gSDOL82PJlfYpM5Wd31KZEFJHrjKJySxla3gWRD94WzDOL5OOSEzlkZiCPEtQw4RYnvBfnHD4/fGfZ9UMI54c7vgr9/78amF05CUanqqShglTWRygZ9nOQ9EbtYpp5jMQirQCMezrZvMRzgqTAobPKgDOtMf1ZAl7b6y+02ksZ+Ms1QMRHgMsLEoltI8LQMVuBfhGhzGaUAEflUBPEL2zjEIF10lLxQbR32GUJ1QIskV9lxbu+9g6sgmKR3N8H3R//O0eeMpaST3u0FPL82wvy/boToHuJHwo8HjvbectehCN9Sr73CUvFc6NUVBu1tz+qwMXbTyyb6qjMDS0of/LapO5iJC5wuyIJFqO4+kZxFK9Ix18EeL6p+2V6qxGlbbitBtyEsPixSDdIxQtzyFWKE0MP/j7ZX4Q+0Vnzo2FAdQ== matt@gradishar.local\n"}]}}, "servicePrincipalProfile": {"clientId": "a0f80ecf-5b3f-4c53-b2b2-3a261f98c19d", "secret": "0af312912e81066301e6"}}}
msrest.http_logger : Response status: 400
msrest.http_logger : Response headers:
msrest.http_logger :     'Cache-Control': 'no-cache'
msrest.http_logger :     'Pragma': 'no-cache'
msrest.http_logger :     'Content-Length': '146'
msrest.http_logger :     'Content-Type': 'application/json'
msrest.http_logger :     'Expires': '-1'
msrest.http_logger :     'x-ms-ratelimit-remaining-subscription-writes': '1196'
msrest.http_logger :     'x-ms-correlation-request-id': 'f562fc6e-7e57-4f17-a888-73b0003d65d2'
msrest.http_logger :     'x-ms-request-id': 'c6e72c6e-9ebc-4940-b08e-dfa3b07ead3f'
msrest.http_logger :     'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'
msrest.http_logger :     'Server': 'nginx'
msrest.http_logger :     'x-ms-routing-request-id': 'WESTUS2:20180511T165613Z:f562fc6e-7e57-4f17-a888-73b0003d65d2'
msrest.http_logger :     'X-Content-Type-Options': 'nosniff'
msrest.http_logger :     'Date': 'Fri, 11 May 2018 16:56:12 GMT'
msrest.http_logger : Response content:
msrest.http_logger : b'{\n  "code": "SubscriptionNotRegistered",\n  "message": "PUT container service operation is not permitted for subscription in Unregistered state"\n }'
msrest.exceptions : Operation failed with status: 'Bad Request'. Details: PUT container service operation is not permitted for subscription in Unregistered state
Operation failed with status: 'Bad Request'. Details: PUT container service operation is not permitted for subscription in Unregistered state
$
```
Maybe there's some important difference between `Unregistered` and `NotRegistered` here?

Reply:
Ah, i see now, this is not a regular ARM error we have expected which is supposed to have the following format to explicitly call out the missing providers:
```
{
  "error": {
    "code": "MissingSubscriptionRegistration",
    "message": "The subscription registration is in 'Unregistered' state. The subscription must be registered to use namespace 'Microsoft.KeyVault'. See https://aka.ms/rps-not-found for how to register subscriptions.",
    "details": [
      {
        "code": "MissingSubscriptionRegistration",
        "target": "Microsoft.KeyVault",
        "message": "The subscription registration is in 'Unregistered' state. The subscription must be registered to use namespace 'Microsoft.KeyVault'. See https://aka.ms/rps-not-found for how to register subscriptions."
      }
    ]
  }
}
```
Maybe a bit of tall order, but can AKS follow the same error pattern? Take out of the context of CLI, this error is no good regardless to any clients, because it is not actionable.

Reply:
> can AKS follow the same error pattern?

Aha, I see. Yes, I think we can modify the error response--that seems like the proper fix. I'll create a work item to do so.

I think we can live with this bug until that's fixed server-side, so unless someone objects I'll close this PR and extract the test changes into a separate one.

Reply:
@mboersma, sounds good. Feel free to let me know if you have any questions when update the service error response. It is possible the `target` field might become a list (say comma separate) for you case. If yes, I can align our logics through the [work item](https://github.com/Azure/msrestazure-for-python/issues/53) at our end.